server/service/transport_error.go

```diff
@@ -33,66 +33,75 @@ func baseError(err string) []map[string]string {
 	}
 }
 
+type validationErrorInterface interface {
+	error
+	Invalid() []map[string]string
+}
+
+type permissionErrorInterface interface {
+	PermissionError() []map[string]string
+}
+
+type notFoundErrorInterface interface {
+	error
+	IsNotFound() bool
+}
+
+type existsErrorInterface interface {
+	error
+	IsExists() bool
+}
+
+type causerInterface interface {
+	Cause() error
+}
+
 // encode error and status header to the client
 func encodeError(ctx context.Context, err error, w http.ResponseWriter) {
 	enc := json.NewEncoder(w)
 	enc.SetIndent("", "  ")
 
-	type validationError interface {
-		error
-		Invalid() []map[string]string
-	}
-	if e, ok := err.(validationError); ok {
+	err = getRootCause(err)
+
+	// TODO: handle the following mysql errors:
+	//	case driver.ErrBadConn:
+	//	case mysql.ErrInvalidConn:
+	//	case mysql.ErrMalformPkt:
+	//	case mysql.ErrNoTLS:
+	//	case mysql.ErrCleartextPassword:
+	//	case mysql.ErrNativePassword:
+	//	case mysql.ErrOldPassword:
+	//	case mysql.ErrUnknownPlugin:
+	//	case mysql.ErrOldProtocol:
+	//	case mysql.ErrPktSync:
+	//	case mysql.ErrPktSyncMul:
+	//	case mysql.ErrPktTooLarge:
+	//	case mysql.ErrBusyBuffer:
```

Comment:
I would be okay with letting these 500 and return the error from MySQL. What do you think?

Reply:
Yeah, I suppose that wouldn't be too bad. They do 500 currently, as they fall within the `default` in the switch. I'll just remove the TODO then.